### ISSUE TITLE
Allow overrides to update-check job spec

### ISSUE THREAD

**Issue by mjnagel** [mattermost/mattermost-operator#290]
#### Summary

From a quick scan through the code I don't believe there is currently any way to override parts of the spec for the update-check job that runs with version updates. It may be desirable to override various parts of this job, for example:
- securityContext: Enforce best practices for security such as runAsNonRoot, dropping all capabilities, etc
- labels/annotations: To better group things together for ownership and provide for integration with externally managed resources (i.e. networkPolicy pod selectors)
- startup command: This is a very niche use-case but when deploying in an istio-injected environment the sidecar needs to be terminated by running additional commands against its endpoint (can provide more details if needed)

Ultimate it would be great to be able to override different portions of the pod spec.

#### Possible fixes

Add a new section to the Mattermost CRD spec that allows for update check pod spec overrides. This would honestly be beneficial for the primary Mattermost pods as well, since currently you have to utilize `resourcePatch` to add labels/annotations to the pods (which is not very beginner friendly or easy to navigate, especially with multiple changes.

Proposed Change:
```yaml
spec:
  podTemplate:
    <allow anything for podspec, or allow subset including annotations/labels/securitycontext
  updateCheck:
    podTemplate:
      <allow anything for podspec, or allow subset including annotations/labels/securitycontext
```

**Comment by mjnagel** [comment]
I'm happy to attempt work on these changes, wanted to open the issue to get the conversation started and see if this would be supported. The primary motivation for this is coming from my work on istio injecting Mattermost, without an ability to disable the sidecar (via labels) or terminate the sidecar (via command override) for the update-check upgrades will require manual intervention to successfully complete.

**Comment by Szymongib** [comment]
Hey @mjnagel, thanks for submitting the issue.

I agree that adding a way to configure an update job would be very useful, as well as other things in the `podTemplate` section. We though about something like that in the past, however there are some drawback that make me a little hesitant for allowing overrides of the whole spec.

A big chunk of `podTemplate` is created based on other things in Mattermost CR or some Operator defaults. This poses a problem, what to do when the user provides his own template? We could just use the provided one and disregard all things that would be set by the Operator, but then user would need to specify all init containers, port, env vars etc. by hand so it seems like using Operator make little sense in such case. We could alternatively do some kind of merging, but then it is also not clear what to do if user explicitly wanted to omit some configuration set by the Operator. We might also run into some configuration edge cases given that some things in Pod Spec can be configured from other CR fields and this information may be used in other places in Operator.

So for exposing the whole `podSpec` in CR, it does not seem to me there is a clear way of doing that without subverting user expectations in some way but still making Operator usable. I am however open to ideas, given that we are thinking of releasing new CR version.

**Comment by mjnagel** [comment]
@Szymongib definitely bring up some good points here...it's a balance between providing customization to the end user and making sure that things won't break with that customization.

Could potentially go the route of throwing out any conflicting overrides (i.e. if I try to override the `app` label, the operator ignores my override). Obviously this would need to be well documented/logged so the end user understands what is happening. The option to override that label would still be available, but would require using `resourcePatch`. Or instead of throwing out conflicts you could have a toggle in the spec that would change the precedence of user values over operator defaults.

Definitely just throwing things out there...I might have to do some research into what other operators have done to handle these scenarios, since I know we make use of several where labels/annotations values are accessible.

**Comment by Szymongib** [comment]
Thanks a lot for diving into this, @mjnagel, and for such detailed write up.

I definitely agree that we can improve with labels and annotations handling as a lot of this things were implemented very early on on never revisited, hence a weird behaviour with labels etc.

We wanted to add support for `securityContext` for some time now as it was requested a bunch of times already. There was even an issue, tho it was closed by the creator #286, I have reopened it as I think it is still something we want to add natively, without the need of using `resourcePatch`. This would be a good starting point to opening customization of Mattermost deployment/pods.

Another thing you mentioned, that I strongly agree with are annotations, I like the idea of not overriding those that should not be overriden, perhaps we could also explore disabling them with something like `monitoring: disabled`, tho that is probably something for the future.

Regarding the labels, `resourceLabels` currently apply to all resources created by Operator. Perhaps we could use more granularity there on top of that. I think what you propose makes total sense, however changing how the labels priority is handled could be considered a breaking change, therefore we would need to postpone releasing it to the next major version.

With the update check job, the spec is copied as this is a simple way to ensure that:
- It avoids a bunch of edge cases of some specific Mattermost configuration.
- It runs with the same config as Mattermost pods, therefore should not do anything unexpected.

I think we should allow some customization there as now it is pretty highly opinionated thing that users should have a way to customize / disable. I would be really cautious with changing the logic of copying the spec, so perhaps we could start with applying some overrides on top of that, and `annotations` seems like a good start point, given that `securityContext` could be copied from the pod spec.


Regarding the shape of the config, we are trying to move more towards grouping things by resource as it makes it easier for user to find and understand particular config options, handle things from code perspective, as well as make it more extendable for the future. I could see therefore something like:
```yaml
spec:
  pod/deployment: # depending how granular we want to go we could do deployment.podTemplate etc...
    podSecurityContext: ...
    annotations: ... # if we want something pod/deployment 
    ...
  updateCheck:
    ...
```
